File: go/0316_remove_duplicate_letters.go
func removeDuplicateLetters(s string) string {
	hashMap := [26]int{}
	visited := [26]bool{}

	for i := range s {
		hashMap[int(s[i])-int('a')] = i
	}

	var res []byte
	for i := range s {
		ch := s[i]
		for len(res) > 0 {
			lastCh := res[len(res)-1]
			if ch <= lastCh && hashMap[int(lastCh)-int('a')] >= i && !visited[int(ch)-int('a')] {
				visited[int(lastCh)-int('a')] = false
				res = res[:len(res)-1]
				continue
			}
			break
		}

		if !visited[int(ch)-int('a')] {
			res = append(res, ch)
			visited[int(ch)-int('a')] = true
		}
	}
	return string(res)
}

File: python/0316_remove_duplicate_letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        def idx(ch):
            return ord(ch) - ord('a')

        counter = [0] * 26
        visited = [False] * 26
        for ch in s:
            counter[idx(ch)] += 1

        stack = []
        for ch in s:
            if visited[idx(ch)]:
                counter[idx(ch)] -= 1
                continue

            while stack and ch < stack[-1] and counter[idx(stack[-1])] > 0:
                visited[idx(stack[-1])] = 0
                stack.pop()

            stack.append(ch)
            counter[idx(ch)] -= 1
            visited[idx(ch)] = True

        return ''.join(stack)
